app/models/voxelytics/VoxelyticsService.scala

```diff
@@ -118,10 +150,41 @@ class VoxelyticsService @Inject()(voxelyticsDAO: VoxelyticsDAO)(implicit ec: Exe
       runUserId <- voxelyticsDAO.getUserIdForRunOpt(runName, user._organization)
     } yield bool2Fox(user.isAdmin || runUserId.forall(_ == user._id))
 
-  def runPublicWrites(run: RunEntry, tasks: List[TaskRunEntry]): JsObject =
-    Json.toJson(run).as[JsObject] ++ Json.obj(
-      "tasks" -> tasks.map(Json.toJson(_))
-    )
+  def taskRunsPublicWrites(combinedTaskRuns: List[CombinedTaskRunEntry], taskRuns: List[TaskRunEntry]): JsArray = {
+    val groupedTaskRuns = taskRuns.groupBy(_.taskName)
+    JsArray(
+      groupedTaskRuns
+        .map(group => {
+          val sortedTaskRuns = group._2.sortBy(_.beginTime).reverse
+          val combinedTaskRun = combinedTaskRuns.find(_.taskName == group._1)
+          Json.obj(
+            "taskName" -> group._1,
+            "state" -> sortedTaskRuns
+              .map(_.state)
+              .find(s => s != VoxelyticsRunState.SKIPPED && s != VoxelyticsRunState.PENDING)
+              .orElse(sortedTaskRuns.headOption.map(_.state))
+              .getOrElse(VoxelyticsRunState.SKIPPED)
+              .asInstanceOf[VoxelyticsRunState],
```

Comment:
Could you add a comment explaining why a cast is needed? (Or better yet, avoid it?)

Reply:
Maybe you can help me here. If I leave this out, the compiler says that it can't figure out the type. If I put the same expression in a `val` and then in the `Json.obj(...)` it works, aswell.

Reply:
Huh, I can reproduce the behavior. This actually smells a bit like a bug in the scala compiler to me. I would very much prefer the `val` variant over `asInstanceOf` though, if that’s ok with you.